### DISCUSSION TITLE
Imperative navigation outside of React.

### DISCUSSION THREAD

**Discussion by madsbuch** [expo/router#324] [General]
It seems like the only way to imperatively navigate is by using the hooks provided (ie. `useRouter`).

How is it recommended to navigate outside of the react context?

**Comment by chmac** [comment]
I also had the same question, and even digging into the code, I couldn't see a way to achieve this. I suspect the reason is because the router needs some context from React in order to work.

I figured a couple of options would work:

- Call `userRouter()` in the root `_layout` and then assign the returned `router` to some global variable
  - Now you can access the router from elsewhere in your codebase
  - But of course it's fragile, there's a risk that your global variable will be `undefined`
- Call `useRouter()` in a component and then pass it down
  - This is what I did in order to be able to use the router inside a thunk

**Comment by wcastand** [comment]
in expo-router v2 you should be able to do so.

https://blog.expo.dev/introducing-expo-router-v2-3850fd5c3ca1

```The new router export can be used as an imperative API to navigate from global stores and providers.```
